lib/uwb/protocols/fira/UwbRegulatoryInformation.cxx

```diff
@@ -0,0 +1,181 @@
+                .SetValue(notstd::to_underlying(Source))
+                .Build())
+        // OUTDOOR_PERMITTED
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::OutdoorPermitted))
+                .SetValue(!!OutdoorPermitted)
+                .Build())
+        // COUNTRY_CODE
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::CountryCode))
+                .SetValue(GetBytesBigEndianFromBitMap(CountryCode, sizeof CountryCode))
+                .Build())
+        // TIMESTAMP
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Timestamp))
+                .SetValue(GetBytesBigEndianFromBitMap(Timestamp, sizeof Timestamp))
+                .Build())
+        // CHANNEL5
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel5))
+                .SetValue(MaximumTransmissionPower.at(Channel::C5))
+                .Build())
+        // CHANNEL6
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel6))
+                .SetValue(MaximumTransmissionPower.at(Channel::C6))
+                .Build())
+        // CHANNEL8
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel8))
+                .SetValue(MaximumTransmissionPower.at(Channel::C8))
+                .Build())
+        // CHANNEL9
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel9))
+                .SetValue(MaximumTransmissionPower.at(Channel::C9))
+                .Build())
+        // CHANNEL10
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel10))
+                .SetValue(MaximumTransmissionPower.at(Channel::C10))
+                .Build())
+        // CHANNEL12
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel12))
+                .SetValue(MaximumTransmissionPower.at(Channel::C12))
+                .Build())
+        // CHANNEL13
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel13))
+                .SetValue(MaximumTransmissionPower.at(Channel::C13))
+                .Build())
+        // CHANNEL14
+        .AddTlv(
+            TlvBer::Builder()
+                .SetTag(notstd::to_underlying(ParameterTag::Channel14))
+                .SetValue(MaximumTransmissionPower.at(Channel::C14))
+                .Build())
+        .Build();
+
+    return tlvBerResult;
+}
+
+/* static */
+UwbRegulatoryInformation
+UwbRegulatoryInformation::FromDataObject(const encoding::TlvBer& tlvBer)

```

Comment:
Should all of these FromDataObject functions signify to the caller if it couldn't decode the tlv? Right now it looks like it returns a default-constructed object if it couldn't decode.

Reply:
That's a good point. The cleanest is probably to `throw` if decoding failed. We'll have to think of the best way to determine when failure actually occurs. These functions seem to need more logic at the end to determine if all required tags were present and decoded properly and if not, then indicate the failure to the caller in some way.

Reply:
I added a basic validation check and if not met, throw `UwbException(UwbStatusGeneric::SyntaxError)`. I think this will bne good enough for now.